operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/Controller.java

```diff
@@ -16,36 +22,98 @@
 import io.javaoperatorsdk.operator.api.config.BaseConfigurationService;
 import io.javaoperatorsdk.operator.api.config.ConfigurationService;
 import io.javaoperatorsdk.operator.api.config.ControllerConfiguration;
+import io.javaoperatorsdk.operator.api.config.DependentResource;
 import io.javaoperatorsdk.operator.api.config.Version;
 import io.javaoperatorsdk.operator.api.monitoring.Metrics;
 import io.javaoperatorsdk.operator.api.monitoring.Metrics.ControllerExecution;
 import io.javaoperatorsdk.operator.api.reconciler.Context;
+import io.javaoperatorsdk.operator.api.reconciler.ContextInitializer;
 import io.javaoperatorsdk.operator.api.reconciler.DeleteControl;
 import io.javaoperatorsdk.operator.api.reconciler.EventSourceContext;
+import io.javaoperatorsdk.operator.api.reconciler.EventSourceContextInjector;
 import io.javaoperatorsdk.operator.api.reconciler.EventSourceInitializer;
 import io.javaoperatorsdk.operator.api.reconciler.Reconciler;
 import io.javaoperatorsdk.operator.api.reconciler.UpdateControl;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.DependentResourceController;
+import io.javaoperatorsdk.operator.api.reconciler.dependent.KubernetesDependentResourceController;
 import io.javaoperatorsdk.operator.processing.event.EventSourceManager;
 import io.javaoperatorsdk.operator.processing.event.source.EventSource;
 
+@SuppressWarnings({"rawtypes", "unchecked"})
 public class Controller<R extends HasMetadata> implements Reconciler<R>,
     LifecycleAware, EventSourceInitializer<R> {
+
+  private static final Logger log = LoggerFactory.getLogger(Controller.class);
+
   private final Reconciler<R> reconciler;
   private final ControllerConfiguration<R> configuration;
   private final KubernetesClient kubernetesClient;
-  private EventSourceManager<R> eventSourceManager;
-  private volatile ConfigurationService configurationService;
+  private final AtomicBoolean started = new AtomicBoolean(false);
+  private List<DependentResourceController> dependents;
+  private final EventSourceManager<R> eventSourceManager;
+  private ConfigurationService configurationService;
 
   public Controller(Reconciler<R> reconciler,
       ControllerConfiguration<R> configuration,
       KubernetesClient kubernetesClient) {
     this.reconciler = reconciler;
     this.configuration = configuration;
     this.kubernetesClient = kubernetesClient;
+
+    eventSourceManager = new EventSourceManager<>(this);
+    final var context = new EventSourceContext<>(
+        eventSourceManager.getControllerResourceEventSource().getResourceCache(),
+        configurationService(), kubernetesClient);
+
+    if (reconciler instanceof EventSourceContextInjector) {
+      EventSourceContextInjector injector = (EventSourceContextInjector) reconciler;
+      injector.injectInto(context);
+    }
+
+    prepareEventSources(context).forEach(eventSourceManager::registerEventSource);
+  }
+
+  private void waitUntilStartedAndInitContext(R resource, Context context) {
```

Comment:
This is to prevent the informers from sending events before everything is ready. Basically, we need all event sources to be ready before events start getting processed.

Reply:
No, that wasn't working with the previous version of the code. You needed that loop because event sources were registered in the constructor before the event manager was started. I moved the registration part back to the start method and that seems to have addressed the issue. I don't remember exactly why the wait was needed but things were not working properly without it.